src/Sarif.Multitool/Rules/SARIF2001.AuthorHighQualityMessages.cs

```diff
@@ -3,85 +3,106 @@
 
 using System;
 using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
 using Microsoft.Json.Pointer;
 
 namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
 {
     public class AuthorHighQualityMessages : SarifValidationSkimmerBase
     {
-        public override MultiformatMessageString FullDescription => new MultiformatMessageString
-        {
-            Text = RuleResources.SARIF2001_AuthorHighQualityMessages_FullDescription_Text
+        /// <summary>
+        /// SARIF2001
+        /// </summary>
+        public override string Id => RuleId.AuthorHighQualityMessages;
+
+        /// <summary>
+        /// Placeholder (full description).
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2001_AuthorHighQualityMessages_FullDescription_Text };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_EnquoteDynamicContent_Text),
+            nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_IncludeDynamicContent_Text),
+            nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_TerminateWithPeriod_Text)
         };
 
         public override FailureLevel DefaultLevel => FailureLevel.Warning;
 
-        public override string Id => RuleId.AuthorHighQualityMessages;
+        private static readonly Regex s_dynamicContentRegex = new Regex(@"{[0-9]+}", RegexOptions.Compiled | RegexOptions.CultureInvariant);
+        private static readonly Regex s_enquoteDynamicContentRegex = new Regex(@"'{[0-9]+}'", RegexOptions.Compiled | RegexOptions.CultureInvariant);
 
-        protected override IEnumerable<string> MessageResourceNames
+        protected override void Analyze(Tool tool, string toolPointer)
         {
-            get
+            if (tool.Driver != null)
             {
-                return new string[]
-                {
-                    nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_TerminateWithPeriod_Text)
-                };
+                AnalyzeToolDriver(tool.Driver, toolPointer.AtProperty(SarifPropertyName.Driver));
             }
         }
 
-        protected override void Analyze(ReportingDescriptor reportingDescriptor, string reportingDescriptorPointer)
+        private void AnalyzeToolDriver(ToolComponent toolComponent, string toolDriverPointer)
         {
-            AnalyzeMessageStrings(reportingDescriptor.MessageStrings, reportingDescriptorPointer);
+            if (toolComponent.Rules != null && toolComponent.Rules.Count > 0)
+            {
+                foreach (ReportingDescriptor rule in toolComponent.Rules)
+                {
+                    AnalyzeReportingDescriptor(rule, toolDriverPointer.AtProperty(SarifPropertyName.Rules));
```

Comment:
>toolDriverPointer.AtProperty(SarifPropertyName.Rules) [](start = 53, length = 53)

You actually want to point to the specific rule, say, the rule at index 0. To do that:

1. Before the `foreach` loop, define `string rulesPointer = toolDriverPointer.AtProperty(SarifPropertyName.Rules;`
2. Inside the `foreach` loop, keep a counter (or just use a simple `for` loop. Suppose the counter is `index`.
3. Then when you call `AnalyzeReportingDescriptor`, the pointer argument should be `rulesPointer.AtIndex(index)`.

Reply:
The result will be `/runs/0/tool/driver/rules/0` (for the rule at index 0) and so on.

---
In reply to: [445100702](https://github.com/microsoft/sarif-sdk/pull/1929#discussion_r445100702) [](ancestors = 445100702)

Reply:
ok. just changed to a `for`, so i dont need to work with the int variable.

---
In reply to: [445101027](https://github.com/microsoft/sarif-sdk/pull/1929#discussion_r445101027) [](ancestors = 445101027,445100702)